Sources/SwiftSyntax/SyntaxData.swift

```diff
@@ -327,43 +331,52 @@ struct SyntaxData {
   ///   - index: The index pointing to where in the raw layout to place this
   ///            child.
   ///   - newChild: The raw syntax for the new child to replace.
+  ///   - newChildArena: The arena in which `newChild` resides.
   ///   - arena: The arena in which the new node will be allocated.
   /// - Returns: The new root node created by this operation, and the new child
   ///            syntax data.
   /// - SeeAlso: replacingSelf(_:)
-  func replacingChild(at index: Int, with newChild: RawSyntax?, arena: SyntaxArena) -> SyntaxData {
-    let newRaw = raw.layoutView!.replacingChild(at: index, with: newChild, arena: arena)
-    return replacingSelf(newRaw, arena: arena)
+  func replacingChild(at index: Int, with newChild: RawSyntax?, rawNodeArena: SyntaxArena?, allocationArena: SyntaxArena) -> SyntaxData {
+    if let newChild {
+      precondition(newChild.arena === rawNodeArena)
```

Comment:
Can this not just be `newChild?.arena == rawNodeArena`?

Reply:
Changed to 
```
precondition(newChild?.arena === rawNodeArena || newChild == nil)
```

We don’t hit the `newChild == nil && rawNodeArena != nil` case at the moment AFAICT but there’s nothing really wrong with that, so I wanted to allow it.

Reply:
Seems odd to me to allow it, they're supposed to be the same so if `newChild` is `nil` and `rawNodeArena` isn't that would suggest something strange is going on.